cpp/include/cudf/join.hpp

```diff
@@ -0,0 +1,381 @@
+    bool                     only_left_in_output;  ///< if true, only output the left column
+  }
+  
+  /** 
+   * @brief  Performs a left non-equi inner join on two tables (left, right)
+   *
+   * A left non-equi inner join acts like an inner join, however it allows
+   * for comparison conditions beyond equality.
+   *
+   * Ultimately, this interface could be quite complex.  For the moment it is limited
+   * to defining a list of conditions that will be anded together to provide the join
+   * criteria.
+   *
+   * @example TableA a: {0, 1, 2}
+   *          TableB b: {1, 2, 3}, a: {1, 2, 5}
+   *          join_ops: { {LESS_THAN, 0, 1, true} }
+   * Result: { a: {1, 2, 5}, b: {1, 2, 3} }
+   *
+   * @example TableA a: {0, 1, 2}, c: {1, 2, 5}
+   *          TableB b: {1, 2, 3}
+   *          join_ops: { {GREATER_THAN, 0, 0, false} }
+   * Result: { a: {2}, b: {1}, c: {5} }
+   *
+   * @throws cudf::logic_error if either table is empty
+   *
+   * @param[in] left               The left table
+   * @param[in] right              The right table
+   * @param[in] join_ops           A vector of join operations.  Each element in the vector
+   *                               defines a pair of columns and the comparison operator
+   *                               that should be applied and a flag indicating whether
+   *                               to include just the column from left, or both columns
+   *                               in the generated output.
+   * @param[in] mr                 Device memory resource to use for device memory allocation
+   *
+   * @returns                      Result of joining `left` and `right` tables on the columns
+   *                               specified by join_ops.  The resulting table will contain
+   *                               all columns of `left` and columns of `right` except for
+   *                               columns from join_ops that are marked to not be included.
+   */
+  std::unique_ptr<cudf::experimental::table> inner_non_equi_join(cudf::table_view const& left,
+                                                                 cudf::table_view const& right,
+                                                                 std::vector<join_operation> const& join_ops,
+                                                                 std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                                                 rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  // TODO:  Think about this.  Should the public API for all joins be this, and have the inner workings call
+  //        the existing API for equi join and the new internal mechanism for non-equi join?  Perhaps
+  //        I implement this public API toward that purpose and we can decide whether to deprecate the
+  //        old interface?
+  /** 
```

Comment:
If these functions are supersets of inner/outer join, I would think this would replace their implementations entirely.  So their interfaces in this header might still exist, but all they'd be doing is calling into this function with an EQUAL operation.    Caveat : I am in no way an expert on database stuff so this comment may be moot :)

Reply:
We'll have to see how the implementations unfold.  I was actually imagining the opposite... that perhaps the interfaces would combine but if the ONLY comparison operator is EQUAL then we would call the existing implementations.  It is possible when all is said and done that this implementation would be comparable in performance to the original, but I'm not convinced at this point.